### PR TITLE
Admins can CRUD a list of Genres

### PR DESCRIPTION
**Pull request by stephaniewillhide**
#29

### CODE REVIEW

File: app/controllers/genres_controller.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def new
+    @genre = Genre.new
+  end
+
+  def create
+    @genre = Genre.new(object_params)
+    if @genre.save
+      redirect_to(genres_path, notice: "#{ @genre.name } successfully created.")
```
Comment (jordanmaguire) [inline app/controllers/genres_controller.rb:20]: 👍🏻  Good use of notices in these actions

File: app/models/genre.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Genre < ApplicationRecord
+  has_and_belongs_to_many :books
+
+  validates :name, presence: true
+
+  scope :ordered, -> { order(name: :asc) }
```
Comment (jordanmaguire) [inline app/models/genre.rb:6]: Could you update this ordering so it is case insensitive? Like in Book in your other PR

File: app/models/genre.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Genre < ApplicationRecord
```
Comment (jordanmaguire) [inline app/models/genre.rb:1]: At the moment we can create Genres with the same name. Does that seem reasonable to you?

File: app/models/genre.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  has_and_belongs_to_many :books
+
+  validates :name, presence: true
+
+  scope :ordered, -> { order(name: :asc) }
+
+  scope :search, -> (search_term) {
+    search_term_with_wildcards = "%#{search_term}%"
+    where("name LIKE ? OR name LIKE ?",
+    search_term_with_wildcards, search_term_with_wildcards) }
```
Comment (jordanmaguire) [inline app/models/genre.rb:11]: When we have blocks that span multiple lines it is customary in Ruby to use the `do` `end` syntax

```ruby
  scope :search, -> (search_term) do
    search_term_with_wildcards = "%#{search_term}%"
    where("name LIKE ? OR name LIKE ?", search_term_with_wildcards, search_term_with_wildcards)
  end
```

File: app/views/books/_book.html.erb [hunk 5]
```diff
@@ -1,6 +1,7 @@
 <tr>
   <td class="px-3"><%= book.name %></td>
   <td class="px-3"><%= book.isbn %></td>
+  <td class="px-3"><%= book.genres.pluck(:name).join(", ") %></td>
```
Comment (jordanmaguire) [inline app/views/books/_book.html.erb:4]: 👍🏻

File: app/views/books/_form.html.erb [hunk 6]
```diff
@@ -2,6 +2,7 @@
   <%= form.input(:name) %>
   <%= form.input(:isbn, label: "ISBN") %>
   <%= form.input(:featured, label: "Featured?", as: :boolean) %>
+  <%= form.association(:genres, collection: Genre.ordered, as: :check_boxes) %>
```
Comment (jordanmaguire) [inline app/views/books/_form.html.erb:5]: 👍🏻

File: config/routes.rb [hunk 7]
```diff
@@ -4,6 +4,8 @@
   devise_for :users
   resources :users, except: [:destroy, :show]
 
+  resources :genres, except: [:show]
```
Comment (jordanmaguire) [inline config/routes.rb:7]: 👍🏻

File: db/migrate/20200904005346_create_genres.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+class CreateGenres < ActiveRecord::Migration[5.2]
+  def change
+    create_table :genres do |t|
+      t.string :name
```
Comment (jordanmaguire) [inline db/migrate/20200904005346_create_genres.rb:4]: Let's add some database protections to ensure we can't put genres with empty names in the database.

To do that, add `null: false` to this line. EG: `t.string :name, null: false`

You can re-run your migrations with either:

1. `rails db:migrate:rollback` and `rails db:migrate`, or
2. `rails db:migrate:redo`

File: db/migrate/20200904005346_create_genres.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+      t.timestamps
+    end
+
+    create_table :books_genres, id: false do |t|
+      t.belongs_to :book
+      t.belongs_to :genre
```
Comment (jordanmaguire) [inline db/migrate/20200904005346_create_genres.rb:10]: Likewise, we can add null protections here

```ruby
t.belongs_to :book, null: false
t.belongs_to :genre, null: false
```

where possible we should always add null protections, to ensure we aren't allowing trash data to get into our database

File: db/migrate/20200904005346_create_genres.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class CreateGenres < ActiveRecord::Migration[5.2]
+  def change
+    create_table :genres do |t|
+      t.string :name
+      t.timestamps
```
Comment (jordanmaguire) [inline db/migrate/20200904005346_create_genres.rb:5]: The same is true here

```
t.timestamps null: false
```

File: db/seeds.rb [hunk 11]
```diff
@@ -7,12 +7,23 @@
...
+  genre = Genre.find_or_create_by(name: name)
+end
+
+[
+  ["A Clockwork Orange", "0393312836", "db/sample_data/a-clockwork-orange.jpg", "Dystopian Fiction"],
+  ["Crime and Punishment", "0679734503", "db/sample_data/crime-and-punishment.jpg", "Psychological Fiction"],
+  ["Transcending CSS", "0321410971", "db/sample_data/transcending-css.jpg", "Reference"],
+  ["The Very Hungry Caterpillar", "0399226907", "db/sample_data/the-very-hungry-caterpillar.jpg", "Children's Literature"]
+].each do |name, isbn, cover, genre|
+  genre = Genre.find_by(name: genre)
```
Comment (jordanmaguire) [inline db/seeds.rb:24]: Let's solve a minor inefficiency and have some fun with ruby.

As this collection gets larger it is going to be less and less effective to find the genre each iteration.

Can you think of a way to not have to find the genre each loop? Keep in mind we are creating them in the loop above starting on line 10.

File: db/seeds.rb [hunk 12]
```diff
@@ -7,12 +7,23 @@
...
+  "Children's Literature"
+].each do |name|
+  genre = Genre.find_or_create_by(name: name)
+end
+
+[
+  ["A Clockwork Orange", "0393312836", "db/sample_data/a-clockwork-orange.jpg", "Dystopian Fiction"],
+  ["Crime and Punishment", "0679734503", "db/sample_data/crime-and-punishment.jpg", "Psychological Fiction"],
+  ["Transcending CSS", "0321410971", "db/sample_data/transcending-css.jpg", "Reference"],
+  ["The Very Hungry Caterpillar", "0399226907", "db/sample_data/the-very-hungry-caterpillar.jpg", "Children's Literature"]
```
Comment (jordanmaguire) [inline db/seeds.rb:22]: Can we seed a book with more than one genre? We may also like to seed a book without a genre. This will give us a more diverse set of seeds, which is useful for finding edge cases when using your dev environment.

**Review by jordanmaguire — COMMENTED** [review]
Further to what I've commented on, could you get some unit tests for the Genre model?

This will require you to either merge your rspec work into master, and then master into this branch, or you could just cherry-pick the commits that setup Rspec into this branch.

File: app/models/genre.rb (hunk 3 above)
Comment (stephaniewillhide) [reply app/models/genre.rb:1]: Nope! Adding the validation in the next commit.

File: app/controllers/genres_controller.rb [hunk 13]
```diff
@@ -0,0 +1,49 @@
+class GenresController < ApplicationController
+  layout "genres"
+  before_action :find_genre, only: [:edit, :update, :destroy, :toggle_featured]
```
Comment (jordanmaguire) [inline app/controllers/genres_controller.rb:3]: There is an action here that doesn't belong. Can you find it and remove it?

File: app/models/genre.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  validates :name,
+            presence: true,
+            uniqueness: true
+
+  scope :ordered, -> { order('LOWER(genres.name)') }
+
+  scope :search, -> (search_term) do
+    search_term_with_wildcards = "%#{search_term}%"
+    where("name LIKE ?",
+    search_term_with_wildcards)
```
Comment (jordanmaguire) [inline app/models/genre.rb:13]: I think this might reader nicer on one line - I really only try to split things across multiple lines when the line gets long (80+ characters)

```ruby
where("name LIKE ?", search_term_with_wildcards)
```

if I split a long line, I'll try to indent the next lines to make it clear that they belong to the above lines:

```ruby
my_cool_method(
  arg_1,
  arg_2,
  arg_3,
)

# or

my_cool_assignment =
  my_cool_method(
    line_1,
    line_2,
    line_3,
  )
```

File: app/models/genre.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+class Genre < ApplicationRecord
+  has_and_belongs_to_many :books
+
+  validates :name,
+            presence: true,
+            uniqueness: true
```
Comment (jordanmaguire) [inline app/models/genre.rb:6]: Some further information about uniqueness validations for you to know -

This validation can't really prevent duplicates from going into the system. If someone was on the create Genre form and they pressed the submit button to submit their genre 10 times rapidly, it will spawn off 10 requests to create the Genre.

It's likely up to 10 of those will go through and at the time of the Genre checking whether its name is unique, there will be one of the preceding Genres on its way into the DB. Therefore the validation will pass, because at that exact moment in time there is no Genre with the same name in the DB.

With these race conditions, of the 10 requests you could end up with maybe 2-3 Genres with the same name in the DB.

This isn't going to be a common occurrence, but it CAN happen.

Something else to know about uniqueness validations is that they are really slow.

Under the hood they basically do something like `Genre.exists?(name: name)`. For a large table that can take 10s/100s of milliseconds, and will happen for every save of every Genre record. The longer it takes, the more time we have for multiple concurrent requests to put "invalid" data in the database.

So - should we really use uniqueness validations in real production apps? Not really - not if we can afford not to.

But - this is all next-level™ knowledge to go through in more detail a little later.

For now - we can leave this validation here but let's also add a database-level validation that CAN well and truly prevent duplicates.

You'll need to add a unique index to your name field.

In your migration, that'll look something like:

`add_index :genres, :name, unique: true`

In the scenario outlined at the top of this comment, the end user would instead see a 500 error on their form.

What would happen is:

1. Request 1 is submitted
2. Request 2 is submitted a few ms later
3. Request 1 commits Genre with name "History" to the DB
4. Concurrently, Request 2 is checking whether a Genre called "History" is in the DB. There isn't at this exact moment in time

Scenario: Validation and no DB index:
5. Request 2 persists a second Genre called "History" to the DB a few ms after Request 1 done

Scenario: Validation and DB index:
6. Request 2 attempts to persist a second Genre called "History" to the DB a few ms after Request 1 done. The DB transaction to do this fails and rolls back, because the Genre from Request 1 is in the DB and this violates the DB index that was created.
7. The end user sees a 500 error on screen. In the fullness of time, you'd teach your model how to listen for unique index exceptions coming from the DB, and handle them gracefully so instead the user would see a validation error like "name is already taken"

File: spec/models/genre_spec.rb [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+  it "validates that a Genre has a name" do
+    genre = Genre.new
+
+    genre.valid?
+    expect(genre.errors[:name]).to eq(["can't be blank"])
+
+    genre.name = "Science Fiction"
+    genre.valid?
+    expect(genre.errors[:name]).to eq([])
+  end
```
Comment (jordanmaguire) [inline spec/models/genre_spec.rb:56]: Could we add a test for the uniqueness validation on name?

**Review by jordanmaguire — COMMENTED** [review]
So far so good, a few little bits to go

File: app/controllers/genres_controller.rb (hunk 13 above)
Comment (stephaniewillhide) [reply app/controllers/genres_controller.rb:3]: Oh my, it's `:toggle_featured`. Removing in the next commit!

**Review by jordanmaguire — APPROVED** [review]
Alright, ready to rock and roll. Good job 💪🏻
